docs/RFCS/20190619_constraint_conformance_report.md

```diff
@@ -0,0 +1,319 @@
+copacetic, there are some specific cases where the lack of information is
+particularly unfortunate:
+
+
+1. Testing fault tolerance: A scenario people ran into was testing CRDB’s
+resiliency to networking failures. A cluster is configured with 3 Availability
+Zones (AZs) and the data is constrained to require a replica in every AZ (or
+even without constraints, the implicit diversity requirement would drive towards
+the same effect). One would think that an AZ can go away at any time and the
+cluster should survive. But that’s only colloquially true. After an AZ goes away
+and subsequently recovers, another AZ cannot disappear right away without
+availability loss; the 2nd AZ has to stay up until constraint conformance is
+re-established. Otherwise, the 2nd AZ likely has 2 replicas for the ranges that
+were migrated away during the first outage.
+As things stand, it’s entirely unclear how one is supposed to know when it’s
+safe to bring down the 2nd AZ. (Of course, here we’re talking about disaster
+testing, not the use of node decomissioning.)
+1. Partitioning a table and running performance tests that assume the partioning
+is in place. For example, say I want to partition my data between Romania and
+Bulgaria and test that my Bulgarian clients indeed start getting local
+latencies. Well, partitioning is obviously not instantaneous and so it’s
+entirely unclear how long I’m supposed to wait until I can conduct my test.
+Same with lease preferences.
+
+## Proposal
+
+We’ll be introducing three ways for admins to observe different aspects of
+replication state: a cluster-wide constraint conformance report visible in the
+AdminUI, new jobs for tracking the progress of constraint changes, and a
+crdb_internal virtual table that can be queried for similar information to the
+report.
+
+
+### crdb_internal.constraint_violations virtual table
+
+Two new system tables will be introduced:
+```sql
+CREATE TABLE crdb_internal.constraint_violations_zones (
+-- the timestamp when this report was generated. Constant throughout the table.
+generated_at timestamp,
+-- zone is the name of the zone that this report refers to.
+zone string,
+-- constraint identifies which constraint this report refers to.
+constraint string,
+-- message is a free-form message describing the violation.
+message string,
+-- the average rate (in ranges/s or MiB/s) by which this quantity changed since
+-- the previous time the report was generated. A negative rate means things are
+-- improving (i.e. amount of violating data is going down).
+change_rate int,
+PRIMARY KEY (object, constraint),
+);
+```
+The other table will be `crdb_internal.constraint_violations_objects`. It's
+identical to the first one, except it describes the data more granularly: at
+the level of tables/indexes/partitions (named "objects" here) rather than zones.
+
+A background process will refresh the data that powers these tables once per
```

Comment:
How sure are we that once per minute is the right answer here? I was thinking that we would want this data updated near real-time so that DBAs do not take actions that could harm the state of the cluster due to faulty information provided from this metric

Reply:
yeah i was thinking you might say that. I guess we stick with the 1m for now.